### ISSUE TITLE
Minimal Accuracy Gain When Adding Text Modality to a simple Image Pipeline

### ISSUE THREAD

**Issue by rahulvigneswaran** [linzhiqiu/cross_modal_adaptation#20]
**Issue Description**
I've built a simple pipeline where I concatenate image and text features from frozen clip encoders into the same batch, apply L2 normalization, and use cross-entropy for training. However, the difference in accuracy with and without the text modality is only about 0.5%, which is quite minimal.

**What I Tried**
- Extract text and image features from frozen clip encoders.
- Concatenated text and image features into a single batch.
- Applied L2 normalization.
- Used standard cross-entropy loss.

**What I Expected**
- A noticeable improvement in accuracy from using both image and text features, more than just 0.5%.

**Questions**
- Are there any additional steps, techniques, or considerations that significantly boost accuracy when incorporating text features with image features.
- Which hyperparameters (e.g., learning rate, batch size, optimizer choices, temperature) might be especially impactful here?

**Comment by linzhiqiu** [comment]
Hey,

Could you clarify what you mean by concatenate? 

And why do you need to build a new pipeline? For a new dataset or task?

Most information like dataset and hyperparameters are included in this repository already, if you have more concrete questions it will help me understand how to help you

**Comment by rahulvigneswaran** [comment]
Hello,

Thank you for your prompt response.

>Could you please clarify your definition of “concatenate”?

```
features = torch.cat([img_features, txt_features], dim=0)
labels = torch.cat([img_labels, txt_labels], dim=0)
```
This is my entire pipeline. 

>And why do you need to build a new pipeline? For a new dataset or task?

Upon reviewing the paper, the pipeline appears straightforward. I attempted to implement it on a subset of ImageNet for my practice. I utilized the handcrafted text features as per your repository.

>if you have more concrete questions it will help me understand how to help you

I’m wondering whether hyperparameters—like temperature—or any additional steps in the pipeline could have a significant impact on accuracy. Since my current implementation only includes those two lines of code, I’m curious if missing or suboptimal settings might be why I’m seeing only a minimal (around 0.5%) improvement from adding text features."

**Comment by linzhiqiu** [comment]
> Hello,
> 
> 
> 
> Thank you for your prompt response.
> 
> 
> 
> >Could you please clarify your definition of “concatenate”?
> 
> 
> 
> ```
> 
> features = torch.cat([img_features, txt_features], dim=0)
> 
> labels = torch.cat([img_labels, txt_labels], dim=0)
> 
> ```
> 
> This is my entire pipeline. 
> 
> 
> 
> >And why do you need to build a new pipeline? For a new dataset or task?
> 
> 
> 
> Upon reviewing the paper, the pipeline appears straightforward. I attempted to implement it on a subset of ImageNet for my practice. I utilized the handcrafted text features as per your repository.
> 
> 
> 
> >if you have more concrete questions it will help me understand how to help you
> 
> 
> 
> I’m wondering whether hyperparameters—like temperature—or any additional steps in the pipeline could have a significant impact on accuracy. Since my current implementation only includes those two lines of code, I’m curious if missing or suboptimal settings might be why I’m seeing only a minimal (around 0.5%) improvement from adding text features."

Thanks for the additional information. Your code seems to be right, and you are right that hyperparameters matter and you will need a held out validation set to grid search for the optimal setting including learning rate and temperature etc. You can find all hyperparameters in my codebase or in the paper. But the optimal hyperparameters certainly depend on your dataset and shot size.

**Comment by linzhiqiu** [comment]
I think learning rate plays the biggest role. If your shot size is large enough you should also finetune the visual encoder (maybe with a smaller learning rate than the linear classifier)

**Comment by rahulvigneswaran** [comment]
Thank you!
